csrc/preseg_passes/fmin_fmax_promotion.cpp

```diff
@@ -0,0 +1,298 @@
+// [NAN  NAN  NAN  NAN  NAN  NAN] <- GOOD_BROADCAST
+enum class IterDomainStatus {
+  // NONE is the status when hitting untracked ID's.
+  NONE,
+
+  // Reduced statuses
+  BAD_REDUCE,
+  GOOD_REDUCE,
+
+  // Full-size statuses
+  DEFAULT,
+  BAD_BROADCAST,
+  BAD_BROADCAST_DEFAULT,
+  GOOD_BROADCAST,
+};
+
+using IterStatusMap = std::unordered_map<IterDomain*, IterDomainStatus>;
+
+IterDomainStatus BopStatus(IterDomainStatus lhs, IterDomainStatus rhs) {
+  if (lhs == IterDomainStatus::NONE) {
+    return rhs;
+  }
+
+  if (rhs == IterDomainStatus::NONE) {
+    return lhs;
+  }
+
+  if (lhs == IterDomainStatus::GOOD_REDUCE ||
+      rhs == IterDomainStatus::GOOD_REDUCE) {
+    return IterDomainStatus::GOOD_REDUCE;
+  } else if (
+      lhs == IterDomainStatus::BAD_REDUCE &&
+      rhs == IterDomainStatus::BAD_REDUCE) {
+    return IterDomainStatus::BAD_REDUCE;
+  }
+
+  if (lhs == rhs) {
+    return lhs;
+  }
+
+  if (lhs == IterDomainStatus::GOOD_BROADCAST ||
+      rhs == IterDomainStatus::GOOD_BROADCAST) {
+    return IterDomainStatus::GOOD_BROADCAST;
+  }
+
+  // The only remaining cases are combinations of DEFAULT,
+  // BAD_BROADCAST, and BAD_BROADCAST_DEFAULT.
+  return IterDomainStatus::BAD_BROADCAST_DEFAULT;
+}
+
+bool StatusIsBad(IterDomainStatus status) {
+  return status == IterDomainStatus::BAD_REDUCE ||
+      status == IterDomainStatus::BAD_BROADCAST ||
+      status == IterDomainStatus::BAD_BROADCAST_DEFAULT;
+}
+
+bool AnyBadInputs(Expr* expr, IterStatusMap& iterMap) {
+  for (auto input : expr->inputs()) {
+    if (auto* in_tv = dynamic_cast<TensorView*>(input)) {
+      for (IterDomain* id : in_tv->getLogicalDomain()) {
+        IterDomainStatus status = iterMap[id];
+        if (StatusIsBad(status)) {
+          return true;
+        }
+      }
+    }
+  }
+
+  return false;
+}
+
+// Once we identify a target reduction, we perform a downward pass starting from
+// the target's direct input. The pass propagates IterDomainStatus information.
+// At the end, we check all output TV's for bad statuses. If at any point we
+// encounter a node we don't know how to propagate information through, we treat
+// it like to a graph output and fail if it has any incoming bad statuses.
+bool AnalyzeMinMaxOp(ReductionOp* targetRop) {
+  Fusion* fusion = targetRop->fusion();
+
+  FusionGuard fg(fusion);
+  ComputeAtLogicalDomainMap logical_map;
+  logical_map.build(true);
+
+  IterStatusMap iterMap;
+
+  auto* in_tv = targetRop->input(0)->as<TensorView>();
+  for (IterDomain* in_id : in_tv->getLogicalDomain()) {
+    iterMap[in_id] = IterDomainStatus::DEFAULT;
+  }
+
+  auto* out_tv = targetRop->output(0)->as<TensorView>();
+  for (IterDomain* out_id : out_tv->getLogicalDomain()) {
+    if (out_id->isReduction()) {
+      iterMap[out_id] = IterDomainStatus::BAD_REDUCE;
+    } else {
+      iterMap[out_id] = IterDomainStatus::DEFAULT;
+    }
+  }
+
+  auto traversal =
+      StmtSort::getExprsBetween({targetRop->input(0)}, fusion->outputs());
```

Comment:
I think you can probably just look between `targetRop->output(0)` and the outputs. The input should be all good, so nothing to analyze from it.